File: 3618-find-the-original-typed-string-ii/find-the-original-typed-string-ii.go
const MOD int = 1e9 + 7

func possibleStringCount(word string, k int) int {
	n := len(word)
	if n < k {
		return 0
	}
	if n == k {
		return 1
	}

	// Tách các đoạn ký tự giống nhau
	seg := []int{1}
	for i := 1; i < n; i++ {
		if word[i] == word[i-1] {
			seg[len(seg)-1]++
		} else {
			seg = append(seg, 1)
		}
	}
	m := len(seg)

	// Tính tổng số tổ hợp chuỗi gốc
	total := 1
	takeMod := false
	for _, x := range seg {
		total *= x
		if total >= MOD {
			total %= MOD
			takeMod = true
		}
	}
	if total == 1 && !takeMod {
		return 1
	}
	if k <= m {
		return total
	}

	// Đếm số tổ hợp không hợp lệ có độ dài < k
	maxT := k - m - 1
	dp := make([][]int, 2)
	for i := range dp {
		dp[i] = make([]int, maxT+1)
	}
	dp[0][0] = 1

	for j := 0; j < m; j++ {
		s := seg[j]
		prefix := make([]int, maxT+2)
		for i := 0; i <= maxT; i++ {
			prefix[i+1] = (prefix[i] + dp[j%2][i]) % MOD
		}
		for i := 0; i <= maxT; i++ {
			L := max(0, i-(s-1))
			R := i
			dp[(j+1)%2][i] = (prefix[R+1] - prefix[L] + MOD) % MOD
		}
	}

	lessK := 0
	for _, x := range dp[m%2] {
		lessK = (lessK + x) % MOD
	}
	return (total - lessK + MOD) % MOD
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 3618-find-the-original-typed-string-ii/find-the-original-typed-string-ii.java
class Solution {
    static final int MOD = (int) 1e9 + 7;

    public int possibleStringCount(String word, int k) {
        int n = word.length();
        if (n < k)
            return 0; // Không thể có chuỗi gốc nào nếu độ dài nhỏ hơn k
        if (n == k)
            return 1; // Nếu bằng nhau, chỉ có 1 khả năng là chính chuỗi đó

        // Xử lý các đoạn ký tự giống nhau liên tiếp
        List<Integer> seg = new ArrayList<>();
        seg.add(1);
        for (int i = 1; i < n; i++)
            if (word.charAt(i) == word.charAt(i - 1))
                seg.set(seg.size() - 1, seg.get(seg.size() - 1) + 1);
            else
                seg.add(1);

        int m = seg.size();

        // Tính tổng số chuỗi gốc có thể, mỗi đoạn có ít nhất 1 ký tự
        long total = 1;
        boolean takeMod = false;
        for (int x : seg) {
            total *= x;
            if (total >= MOD) {
                total %= MOD;
                takeMod = true;
            }
        }

        if (total == 1 && !takeMod)
            return 1; // Trường hợp đặc biệt: không có đoạn nào bị lặp

        if (k <= m)
            return (int) total; // Nếu k <= số đoạn thì mọi chuỗi gốc đều hợp lệ

        // Đếm số chuỗi gốc có độ dài nhỏ hơn k (invalid strings)
        // Ta cần đếm số cách phân bố số lần lặp lại sao cho tổng z0 + z1 + ... + z[m-1] <= k - m - 1
        // với 0 <= zi <= seg[i] - 1
        int maxT = k - m - 1;

        long[][] dp = new long[2][maxT + 1];
        long[] prefix = new long[maxT + 2]; // prefix[i+1] = tổng dp[j][0..i]
        dp[0][0] = 1; // cơ sở: chỉ có 1 cách để tạo độ dài 0

        for (int j = 0; j < m; j++) {
            int s = seg.get(j);
            // Tính prefix sum cho dp[j % 2]
            prefix[0] = 0;
            for (int i = 0; i <= maxT; i++)
                prefix[i + 1] = (prefix[i] + dp[j % 2][i]) % MOD;

            // Tính dp[(j + 1) % 2][i] sử dụng sliding window prefix
            for (int i = 0; i <= maxT; i++) {
                int L = Math.max(0, i - (s - 1));
                int R = i;
                dp[(j + 1) % 2][i] = (prefix[R + 1] - prefix[L] + MOD) % MOD;
            }
        }

        // Tổng số chuỗi gốc có độ dài < k
        long lessK = 0;
        for (int i = 0; i <= maxT; i++)
            lessK = (lessK + dp[m % 2][i]) % MOD;

        // Trả về total - lessK (modulo)
        return (int) ((total - lessK + MOD) % MOD);
    }
}
